File: PKEYConfigExtractor/keycutter.go/keycutter.go
package main

import (
	"flag"
	"fmt"
	"math/big"
	"os"
	"regexp"
	"strconv"
	"strings"
)

const ALPHABET = "BCDFGHJKMPQRTVWXY2346789"

var CRC32_TABLE []uint32

func init() {
	CRC32_TABLE = make([]uint32, 256)
	for i := 0; i < 256; i++ {
		k := uint32(i) << 24
		for bit := 0; bit < 8; bit++ {
			if k&0x80000000 != 0 {
				k = (k << 1) ^ 0x4c11db7
			} else {
				k = k << 1
			}
		}
		CRC32_TABLE[i] = k & 0xffffffff
	}
}

type ProductKeyDecoder struct {
	Key5x5   string
	Key      *big.Int
	Group    uint64
	Serial   uint64
	Security uint64
	Checksum uint64
	Upgrade  uint64
	Extra    uint64
}

func decode5x5(key string) *big.Int {
	key = strings.ReplaceAll(key, "-", "")
	
	dec := []int{strings.Index(key, "N")}
	keyNoN := strings.ReplaceAll(key, "N", "")
	
	for _, l := range keyNoN {
		dec = append(dec, strings.IndexRune(ALPHABET, l))
	}
	
	result := big.NewInt(0)
	twentyFour := big.NewInt(24)
	
	for _, x := range dec {
		result.Mul(result, twentyFour)
		result.Add(result, big.NewInt(int64(x)))
	}
	
	return result
}

func NewProductKeyDecoder(key string) *ProductKeyDecoder {
	keyInt := decode5x5(key)
	
	mask1 := new(big.Int)
	mask1.SetString("fffff", 16)
	
	mask2 := new(big.Int)
	mask2.SetString("3fffffff00000", 16)
	
	mask3 := new(big.Int)
	mask3.SetString("1fffffffffffff000000000000", 16)
	
	mask4 := new(big.Int)
	mask4.SetString("3ff0000000000000000000000000", 16)
	
	mask5 := new(big.Int)
	mask5.SetString("20000000000000000000000000000", 16)
	
	mask6 := new(big.Int)
	mask6.SetString("40000000000000000000000000000", 16)
	
	group := new(big.Int).And(keyInt, mask1)
	serial := new(big.Int).And(keyInt, mask2)
	serial.Rsh(serial, 20)
	security := new(big.Int).And(keyInt, mask3)
	security.Rsh(security, 50)
	checksum := new(big.Int).And(keyInt, mask4)
	checksum.Rsh(checksum, 103)
	upgrade := new(big.Int).And(keyInt, mask5)
	upgrade.Rsh(upgrade, 113)
	extra := new(big.Int).And(keyInt, mask6)
	extra.Rsh(extra, 114)
	
	return &ProductKeyDecoder{
		Key5x5:   key,
		Key:      keyInt,
		Group:    group.Uint64(),
		Serial:   serial.Uint64(),
		Security: security.Uint64(),
		Checksum: checksum.Uint64(),
		Upgrade:  upgrade.Uint64(),
		Extra:    extra.Uint64(),
	}
}

type ProductKeyEncoder struct {
	Key5x5   string
	Key      *big.Int
	Group    uint64
	Serial   uint64
	Security uint64
	Checksum uint64
	Upgrade  uint64
	Extra    uint64
}

func encode(key *big.Int) []byte {
	twentyFour := big.NewInt(24)
	temp := new(big.Int).Set(key)
	num := big.NewInt(0)
	
	for i := 0; i < 25; i++ {
		mod := new(big.Int)
		temp.DivMod(temp, twentyFour, mod)
		
		num.Lsh(num, 8)
		num.Or(num, mod)
	}

	numBytes := num.Bytes()

	result := make([]byte, 25)
	if len(numBytes) <= 25 {
		copy(result[25-len(numBytes):], numBytes)
	} else {
		copy(result, numBytes[len(numBytes)-25:])
	}
	for i := 0; i < 12; i++ {
		result[i], result[24-i] = result[24-i], result[i]
	}
	
	return result
}

func to5x5(key *big.Int) string {
	keyBytes := encode(key)
	key5x5 := make([]byte, 0, 24)
	for i := 1; i < 25; i++ {
		key5x5 = append(key5x5, ALPHABET[keyBytes[i]])
	}
	nPos := int(keyBytes[0])
	if nPos > len(key5x5) {
		nPos = len(key5x5)
	}
	result := make([]byte, 0, 25)
	result = append(result, key5x5[:nPos]...)
	result = append(result, 'N')
	result = append(result, key5x5[nPos:]...)
	
	final := string(result[:5]) + "-" + string(result[5:10]) + "-" + 
	         string(result[10:15]) + "-" + string(result[15:20]) + "-" + 
	         string(result[20:])
	
	return final
}

func checksumKey(key *big.Int) uint64 {

	keyBytes := key.Bytes()
	
	padded := make([]byte, 16)
	if len(keyBytes) <= 16 {
		copy(padded[16-len(keyBytes):], keyBytes)
	} else {
		copy(padded, keyBytes[len(keyBytes)-16:])
	}
	

	for i := 0; i < 8; i++ {
		padded[i], padded[15-i] = padded[15-i], padded[i]
	}
	
	crc := uint32(0xffffffff)
	for _, b := range padded {
		crc = (crc << 8) ^ CRC32_TABLE[((crc>>24)^uint32(b))&0xff]
	}
	
	return uint64(^crc & 0x3ff)
}

func NewProductKeyEncoder(group, serial, security, upgrade, checksum, extra uint64) (*ProductKeyEncoder, error) {

	if group > 0xfffff || serial > 0x3fffffff || security > 0x1fffffffffffff ||
		checksum > 0x400 || upgrade > 0x1 || extra > 0x1 {
		return nil, fmt.Errorf("key parameter(s) not within bounds")
	}
	key := big.NewInt(0)
	
	if extra > 0 {
		temp := big.NewInt(int64(extra))
		temp.Lsh(temp, 114)
		key.Or(key, temp)
	}
	
	if upgrade > 0 {
		temp := big.NewInt(int64(upgrade))
		temp.Lsh(temp, 113)
		key.Or(key, temp)
	}
	
	if security > 0 {
		temp := big.NewInt(int64(security))
		temp.Lsh(temp, 50)
		key.Or(key, temp)
	}
	
	if serial > 0 {
		temp := big.NewInt(int64(serial))
		temp.Lsh(temp, 20)
		key.Or(key, temp)
	}
	
	if group > 0 {
		temp := big.NewInt(int64(group))
		key.Or(key, temp)
	}
	
	if checksum == 0x400 {
		checksum = checksumKey(key)
	}
	
	temp := big.NewInt(int64(checksum))
	temp.Lsh(temp, 103)
	key.Or(key, temp)
	
	if extra != 0 {
		maxVal := new(big.Int)
		maxVal.SetString("62A32B15518", 16)
		maxVal.Lsh(maxVal, 72)
		if key.Cmp(maxVal) > 0 {
			return nil, fmt.Errorf("extra parameter unencodable")
		}
	}
	
	return &ProductKeyEncoder{
		Key:      key,
		Key5x5:   to5x5(key),
		Group:    group,
		Serial:   serial,
		Security: security,
		Checksum: checksum,
		Upgrade:  upgrade,
		Extra:    extra,
	}, nil
}

func parseHexOrDec(s string) (uint64, error) {
	if strings.HasPrefix(s, "0x") || strings.HasPrefix(s, "0X") {
		return strconv.ParseUint(s[2:], 16, 64)
	}
	return strconv.ParseUint(s, 0, 64)
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: keycutter [encode|decode]")
		os.Exit(1)
	}
	
	command := os.Args[1]
	
	switch command {
	case "decode":
		decodeCmd := flag.NewFlagSet("decode", flag.ExitOnError)
		output := decodeCmd.String("output", "parametric", "Output format: parametric, raw, rawhex")
		decodeCmd.Parse(os.Args[2:])
		
		if decodeCmd.NArg() < 1 {
			fmt.Println("Usage: keycutter decode [-output format] <key>")
			os.Exit(1)
		}
		
		key := decodeCmd.Arg(0)
		

		alphaPattern := fmt.Sprintf("(?:[%sN]{5}-){4}[%sN]{4}[%s]", ALPHABET, ALPHABET, ALPHABET)
		matched, _ := regexp.MatchString(alphaPattern, key)
		if !matched {
			fmt.Println("Invalid product key")
			os.Exit(1)
		}
		
		keyi := NewProductKeyDecoder(key)
		
		switch *output {
		case "parametric":
			fmt.Printf("\nPKey     : [%s]\n        -> [%032x]\n\n", keyi.Key5x5, keyi.Key)
			fmt.Println("            0xfffff")
			fmt.Printf("Group    : [0x%05x]\n\n", keyi.Group)
			fmt.Println("            0x3fffffff")
			fmt.Printf("Serial   : [0x%08x]\n\n", keyi.Serial)
			fmt.Println("            0x1FFFFFFFFFFFFF")
			fmt.Printf("Security : [0x%014x]\n\n", keyi.Security)
			fmt.Println("            0x3FF")
			fmt.Printf("Checksum : [0x%03x]\n\n", keyi.Checksum)
			fmt.Println("            0x1")
			fmt.Printf("Upgrade  : [0x%01x]\n\n", keyi.Upgrade)
			fmt.Println("            0x1")
			fmt.Printf("Extra    : [0x%01x]\n\n", keyi.Extra)
		case "raw":
			fmt.Printf("%s\n%s\n%d\n%d\n%d\n%d\n%d\n%d\n",
				keyi.Key5x5, keyi.Key.String(),
				keyi.Group, keyi.Serial, keyi.Security,
				keyi.Upgrade, keyi.Checksum, keyi.Extra)
		case "rawhex":
			fmt.Printf("%s\n%s\n0x%x\n0x%x\n0x%x\n0x%x\n0x%x\n0x%x\n",
				keyi.Key5x5, keyi.Key.Text(16),
				keyi.Group, keyi.Serial, keyi.Security,
				keyi.Upgrade, keyi.Checksum, keyi.Extra)
		}
		
	case "encode":
		encodeCmd := flag.NewFlagSet("encode", flag.ExitOnError)
		upgrade := encodeCmd.String("u", "0", "Upgrade bit")
		checksum := encodeCmd.String("c", "0x400", "Checksum (0x400 for automatic)")
		extra := encodeCmd.String("e", "0", "Extra bit")
		encodeCmd.Parse(os.Args[2:])
		
		if encodeCmd.NArg() < 3 {
			fmt.Println("Usage: keycutter encode [-u upgrade] [-c checksum] [-e extra] <group> <serial> <security>")
			os.Exit(1)
		}
		
		group, _ := parseHexOrDec(encodeCmd.Arg(0))
		serial, _ := parseHexOrDec(encodeCmd.Arg(1))
		security, _ := parseHexOrDec(encodeCmd.Arg(2))
		upgradeVal, _ := parseHexOrDec(*upgrade)
		checksumVal, _ := parseHexOrDec(*checksum)
		extraVal, _ := parseHexOrDec(*extra)
		
		keyi, err := NewProductKeyEncoder(group, serial, security, upgradeVal, checksumVal, extraVal)
		if err != nil {
			fmt.Println("Error:", err)
			os.Exit(1)
		}
		
		fmt.Println(keyi.Key5x5)
		
	default:
		fmt.Println("Unknown command. Use: encode, decode, or template")
		os.Exit(1)
	}
}

File: PKEYConfigExtractor/keycutter.cpp/keycutter.cpp
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <string>
#include <vector>
#include <algorithm>
#include <regex>
#include <iostream>

using namespace std;

const string ALPHABET = "BCDFGHJKMPQRTVWXY2346789";
uint32_t CRC32_TABLE[256];

void initCrc(){
    for(int i=0;i<256;i++){
        uint32_t k = (uint32_t)i << 24;
        for(int bit=0; bit<8; bit++){
            if(k & 0x80000000u) k = (k<<1) ^ 0x4c11db7u;
            else k = k<<1;
        }
        CRC32_TABLE[i] = k & 0xffffffffu;
    }
}

struct BigInt {
    vector<uint32_t> limbs; // little-endian, base 2^32

    BigInt() {}
    BigInt(uint64_t v){
        if(v){ limbs.push_back((uint32_t)v); if(v>>32) limbs.push_back((uint32_t)(v>>32)); }
    }

    void trim(){ while(!limbs.empty() && limbs.back()==0) limbs.pop_back(); }
    bool isZero() const { return limbs.empty(); }

    static int hexVal(char c){
        if(c>='0'&&c<='9') return c-'0';
        if(c>='a'&&c<='f') return c-'a'+10;
        if(c>='A'&&c<='F') return c-'A'+10;
        return 0;
    }

    static BigInt fromHex(const string& hex){
        BigInt r;
        for(char c: hex){
            r.mulSmall(16);
            r.addSmall((uint64_t)hexVal(c));
        }
        return r;
    }

    void mulSmall(uint64_t m){
        uint64_t carry=0;
        for(auto &l: limbs){
            uint64_t v = (uint64_t)l*m + carry;
            l = (uint32_t)v;
            carry = v>>32;
        }
        while(carry){ limbs.push_back((uint32_t)carry); carry>>=32; }
        trim();
    }

    void addSmall(uint64_t a){
        size_t i=0;
        while(a){
            if(i>=limbs.size()) limbs.push_back(0);
            uint64_t v = (uint64_t)limbs[i] + (a & 0xffffffffULL);
            limbs[i] = (uint32_t)v;
            a = (a>>32) + (v>>32);
            i++;
        }
        trim();
    }

    uint64_t divmodSmall(uint64_t d){
        uint64_t rem=0;
        for(int i=(int)limbs.size()-1;i>=0;i--){
            uint64_t cur = (rem<<32) | (uint64_t)limbs[i];
            limbs[i] = (uint32_t)(cur/d);
            rem = cur % d;
        }
        trim();
        return rem;
    }

    void shl(int n){
        if(isZero()||n==0) return;
        int limbShift = n/32, bitShift = n%32;
        vector<uint32_t> res(limbs.size()+limbShift+1,0);
        for(size_t i=0;i<limbs.size();i++){
            uint64_t v = (uint64_t)limbs[i] << bitShift;
            res[i+limbShift] |= (uint32_t)v;
            res[i+limbShift+1] |= (uint32_t)(v>>32);
        }
        limbs = res; trim();
    }

    void shr(int n){
        if(n==0) return;
        int limbShift = n/32, bitShift = n%32;
        if((size_t)limbShift >= limbs.size()){ limbs.clear(); return; }
        vector<uint32_t> res(limbs.size()-limbShift,0);
        for(size_t i=0;i<res.size();i++){
            uint64_t lo = (uint64_t)limbs[i+limbShift] >> bitShift;
            uint64_t hi = 0;
            if(bitShift>0 && i+limbShift+1<limbs.size())
                hi = (uint64_t)limbs[i+limbShift+1] << (32-bitShift);
            res[i] = (uint32_t)(lo|hi);
        }
        limbs = res; trim();
    }

    BigInt bitAnd(const BigInt& o) const{
        BigInt r; size_t n=min(limbs.size(),o.limbs.size());
        r.limbs.resize(n);
        for(size_t i=0;i<n;i++) r.limbs[i]=limbs[i]&o.limbs[i];
        r.trim(); return r;
    }

    int cmp(const BigInt& o) const{
        if(limbs.size()!=o.limbs.size()) return limbs.size()<o.limbs.size()?-1:1;
        for(int i=(int)limbs.size()-1;i>=0;i--)
            if(limbs[i]!=o.limbs[i]) return limbs[i]<o.limbs[i]?-1:1;
        return 0;
    }

    uint64_t toUint64() const{
        uint64_t r=0;
        if(limbs.size()>0) r = limbs[0];
        if(limbs.size()>1) r |= (uint64_t)limbs[1]<<32;
        return r;
    }

    vector<uint8_t> bytes() const{
        vector<uint8_t> r;
        for(int i=(int)limbs.size()-1;i>=0;i--){
            for(int b=3;b>=0;b--){
                uint8_t byte = (limbs[i]>>(b*8))&0xff;
                if(!r.empty()||byte!=0) r.push_back(byte);
            }
        }
        return r;
    }

    string toHexPadded(int width) const{
        string s;
        for(int i=(int)limbs.size()-1;i>=0;i--){
            char buf[9];
            snprintf(buf,sizeof(buf),"%08x",limbs[i]);
            s+=buf;
        }
        if(s.empty()) s="0";
        while((int)s.size()<width) s="0"+s;
        return s;
    }

    string toHex() const{
        if(isZero()) return "0";
        string s;
        for(int i=(int)limbs.size()-1;i>=0;i--){
            char buf[9];
            if(i==(int)limbs.size()-1) snprintf(buf,sizeof(buf),"%x",limbs[i]);
            else snprintf(buf,sizeof(buf),"%08x",limbs[i]);
            s+=buf;
        }
        return s;
    }

    string toDec() const{
        if(isZero()) return "0";
        BigInt tmp = *this;
        string s;
        while(!tmp.isZero()){
            uint64_t r = tmp.divmodSmall(10);
            s += char('0'+r);
        }
        reverse(s.begin(), s.end());
        return s;
    }
};

// ---------------- Product key decode/encode ----------------
struct ProductKeyDecoder {
    string Key5x5;
    BigInt Key;
    uint64_t Group, Serial, Security, Checksum, Upgrade, Extra;
};

BigInt decode5x5(string key){
    string k;
    for(char c: key) if(c!='-') k+=c;

    int nPos = (int)k.find('N');

    string kNoN;
    for(char c: k) if(c!='N') kNoN+=c;

    vector<int> dec;
    dec.push_back(nPos);
    for(char c: kNoN) dec.push_back((int)ALPHABET.find(c));

    BigInt result;
    for(int x: dec){
        result.mulSmall(24);
        result.addSmall((uint64_t)x);
    }
    return result;
}

ProductKeyDecoder NewProductKeyDecoder(const string& key){
    BigInt keyInt = decode5x5(key);

    BigInt mask1 = BigInt::fromHex("fffff");
    BigInt mask2 = BigInt::fromHex("3fffffff00000");
    BigInt mask3 = BigInt::fromHex("1fffffffffffff000000000000");
    BigInt mask4 = BigInt::fromHex("3ff0000000000000000000000000");
    BigInt mask5 = BigInt::fromHex("20000000000000000000000000000");
    BigInt mask6 = BigInt::fromHex("40000000000000000000000000000");

    BigInt group = keyInt.bitAnd(mask1);

    BigInt serial = keyInt.bitAnd(mask2); serial.shr(20);
    BigInt security = keyInt.bitAnd(mask3); security.shr(50);
    BigInt checksum = keyInt.bitAnd(mask4); checksum.shr(103);
    BigInt upgrade = keyInt.bitAnd(mask5); upgrade.shr(113);
    BigInt extra = keyInt.bitAnd(mask6); extra.shr(114);

    ProductKeyDecoder d;
    d.Key5x5 = key;
    d.Key = keyInt;
    d.Group = group.toUint64();
    d.Serial = serial.toUint64();
    d.Security = security.toUint64();
    d.Checksum = checksum.toUint64();
    d.Upgrade = upgrade.toUint64();
    d.Extra = extra.toUint64();
    return d;
}

vector<uint8_t> encodeBytes(BigInt key){
    BigInt temp = key;
    BigInt num;
    for(int i=0;i<25;i++){
        uint64_t mod = temp.divmodSmall(24);
        num.shl(8);
        num.addSmall(mod);
    }
    vector<uint8_t> numBytes = num.bytes();
    vector<uint8_t> result(25,0);
    if(numBytes.size()<=25){
        size_t offset = 25-numBytes.size();
        for(size_t i=0;i<numBytes.size();i++) result[offset+i]=numBytes[i];
    } else {
        for(int i=0;i<25;i++) result[i]=numBytes[numBytes.size()-25+i];
    }
    for(int i=0;i<12;i++) swap(result[i], result[24-i]);
    return result;
}

string to5x5(BigInt key){
    vector<uint8_t> keyBytes = encodeBytes(key);
    string key5x5;
    for(int i=1;i<25;i++) key5x5 += ALPHABET[keyBytes[i]];

    int nPos = (int)keyBytes[0];
    if(nPos > (int)key5x5.size()) nPos = (int)key5x5.size();

    string result = key5x5.substr(0,nPos) + "N" + key5x5.substr(nPos);

    return result.substr(0,5)+"-"+result.substr(5,5)+"-"+result.substr(10,5)+"-"+
           result.substr(15,5)+"-"+result.substr(20,5);
}

uint64_t checksumKey(BigInt key){
    vector<uint8_t> keyBytes = key.bytes();
    vector<uint8_t> padded(16,0);
    if(keyBytes.size()<=16){
        size_t off = 16-keyBytes.size();
        for(size_t i=0;i<keyBytes.size();i++) padded[off+i]=keyBytes[i];
    } else {
        for(int i=0;i<16;i++) padded[i]=keyBytes[keyBytes.size()-16+i];
    }
    for(int i=0;i<8;i++) swap(padded[i], padded[15-i]);

    uint32_t crc = 0xffffffffu;
    for(uint8_t b: padded){
        crc = (crc<<8) ^ CRC32_TABLE[((crc>>24)^(uint32_t)b)&0xff];
    }
    return (uint64_t)(~crc) & 0x3ffULL;
}

struct ProductKeyEncoder {
    string Key5x5;
    BigInt Key;
    uint64_t Group, Serial, Security, Checksum, Upgrade, Extra;
};

bool NewProductKeyEncoder(uint64_t group, uint64_t serial, uint64_t security,
                           uint64_t upgrade, uint64_t checksum, uint64_t extra,
                           ProductKeyEncoder& out, string& err){
    if(group>0xfffffULL || serial>0x3fffffffULL || security>0x1fffffffffffffULL ||
       checksum>0x400ULL || upgrade>0x1ULL || extra>0x1ULL){
        err = "key parameter(s) not within bounds";
        return false;
    }

    BigInt key;

    if(extra>0){
        BigInt t((uint64_t)extra);
        t.shl(114);
        for(size_t i=0;i<t.limbs.size();i++){
            if(i>=key.limbs.size()) key.limbs.push_back(0);
            key.limbs[i] |= t.limbs[i];
        }
    }
    if(upgrade>0){
        BigInt t((uint64_t)upgrade);
        t.shl(113);
        for(size_t i=0;i<t.limbs.size();i++){
            if(i>=key.limbs.size()) key.limbs.push_back(0);
            key.limbs[i] |= t.limbs[i];
        }
    }
    if(security>0){
        BigInt t((uint64_t)security);
        t.shl(50);
        for(size_t i=0;i<t.limbs.size();i++){
            if(i>=key.limbs.size()) key.limbs.push_back(0);
            key.limbs[i] |= t.limbs[i];
        }
    }
    if(serial>0){
        BigInt t((uint64_t)serial);
        t.shl(20);
        for(size_t i=0;i<t.limbs.size();i++){
            if(i>=key.limbs.size()) key.limbs.push_back(0);
            key.limbs[i] |= t.limbs[i];
        }
    }
    if(group>0){
        BigInt t((uint64_t)group);
        for(size_t i=0;i<t.limbs.size();i++){
            if(i>=key.limbs.size()) key.limbs.push_back(0);
            key.limbs[i] |= t.limbs[i];
        }
    }
    key.trim();

    if(checksum==0x400ULL) checksum = checksumKey(key);

    {
        BigInt t((uint64_t)checksum);
        t.shl(103);
        for(size_t i=0;i<t.limbs.size();i++){
            if(i>=key.limbs.size()) key.limbs.push_back(0);
            key.limbs[i] |= t.limbs[i];
        }
        key.trim();
    }

    if(extra!=0){
        BigInt maxVal = BigInt::fromHex("62A32B15518");
        maxVal.shl(72);
        if(key.cmp(maxVal) > 0){
            err = "extra parameter unencodable";
            return false;
        }
    }

    out.Key = key;
    out.Key5x5 = to5x5(key);
    out.Group = group;
    out.Serial = serial;
    out.Security = security;
    out.Checksum = checksum;
    out.Upgrade = upgrade;
    out.Extra = extra;
    return true;
}

// ---------------- CLI helpers ----------------
bool parseHexOrDec(const string& s, uint64_t& out){
    try{
        if(s.size()>2 && (s[0]=='0') && (s[1]=='x'||s[1]=='X')){
            out = stoull(s.substr(2), nullptr, 16);
        } else {
            out = stoull(s, nullptr, 10);
        }
        return true;
    } catch(...) {
        return false;
    }
}

// crude getopt-like flag extraction: removes "-flag value" or "-flag=value"
// pairs from args and returns remaining positional arguments
vector<string> extractFlags(const vector<string>& args, vector<pair<string,string>>& flagDefs){
    // flagDefs: name -> default; filled with parsed value
    vector<string> positional;
    for(size_t i=0;i<args.size();i++){
        const string& a = args[i];
        bool matched=false;
        if(a.size()>1 && a[0]=='-'){
            string name, val;
            auto eq = a.find('=');
            if(eq!=string::npos){
                name = a.substr(1, eq-1);
                val = a.substr(eq+1);
            } else {
                name = a.substr(1);
                if(i+1<args.size()){ val = args[i+1]; i++; }
            }
            for(auto& fd: flagDefs){
                if(fd.first==name){ fd.second = val; matched=true; break; }
            }
        }
        if(!matched) positional.push_back(a);
    }
    return positional;
}

int main(int argc, char* argv[]){
    initCrc();

    vector<string> args(argv+1, argv+argc);
    if(args.empty()){
        printf("Usage: keycutter [encode|decode]\n");
        return 1;
    }

    string command = args[0];
    vector<string> rest(args.begin()+1, args.end());

    if(command=="decode"){
        vector<pair<string,string>> flags = {{"output","parametric"}};
        vector<string> pos = extractFlags(rest, flags);
        string output = flags[0].second;

        if(pos.empty()){
            printf("Usage: keycutter decode [-output format] <key>\n");
            return 1;
        }
        string key = pos[0];

        string pattern = "(?:[" + ALPHABET + "N]{5}-){4}[" + ALPHABET + "N]{4}[" + ALPHABET + "]";
        std::regex re(pattern);
        if(!std::regex_search(key, re)){
            printf("Invalid product key\n");
            return 1;
        }

        ProductKeyDecoder keyi = NewProductKeyDecoder(key);

        if(output=="parametric"){
            printf("\nPKey     : [%s]\n        -> [%s]\n\n", keyi.Key5x5.c_str(), keyi.Key.toHexPadded(32).c_str());
            printf("            0xfffff\n");
            printf("Group    : [0x%05llx]\n\n", (unsigned long long)keyi.Group);
            printf("            0x3fffffff\n");
            printf("Serial   : [0x%08llx]\n\n", (unsigned long long)keyi.Serial);
            printf("            0x1FFFFFFFFFFFFF\n");
            printf("Security : [0x%014llx]\n\n", (unsigned long long)keyi.Security);
            printf("            0x3FF\n");
            printf("Checksum : [0x%03llx]\n\n", (unsigned long long)keyi.Checksum);
            printf("            0x1\n");
            printf("Upgrade  : [0x%01llx]\n\n", (unsigned long long)keyi.Upgrade);
            printf("            0x1\n");
            printf("Extra    : [0x%01llx]\n\n", (unsigned long long)keyi.Extra);
        } else if(output=="raw"){
            printf("%s\n%s\n%llu\n%llu\n%llu\n%llu\n%llu\n%llu\n",
                keyi.Key5x5.c_str(), keyi.Key.toDec().c_str(),
                (unsigned long long)keyi.Group, (unsigned long long)keyi.Serial,
                (unsigned long long)keyi.Security, (unsigned long long)keyi.Upgrade,
                (unsigned long long)keyi.Checksum, (unsigned long long)keyi.Extra);
        } else if(output=="rawhex"){
            printf("%s\n%s\n0x%llx\n0x%llx\n0x%llx\n0x%llx\n0x%llx\n0x%llx\n",
                keyi.Key5x5.c_str(), keyi.Key.toHex().c_str(),
                (unsigned long long)keyi.Group, (unsigned long long)keyi.Serial,
                (unsigned long long)keyi.Security, (unsigned long long)keyi.Upgrade,
                (unsigned long long)keyi.Checksum, (unsigned long long)keyi.Extra);
        }

    } else if(command=="encode"){
        vector<pair<string,string>> flags = {{"u","0"},{"c","0x400"},{"e","0"}};
        vector<string> pos = extractFlags(rest, flags);

        if(pos.size()<3){
            printf("Usage: keycutter encode [-u upgrade] [-c checksum] [-e extra] <group> <serial> <security>\n");
            return 1;
        }

        uint64_t group=0, serial=0, security=0, upgradeVal=0, checksumVal=0, extraVal=0;
        parseHexOrDec(pos[0], group);
        parseHexOrDec(pos[1], serial);
        parseHexOrDec(pos[2], security);
        parseHexOrDec(flags[0].second, upgradeVal);
        parseHexOrDec(flags[1].second, checksumVal);
        parseHexOrDec(flags[2].second, extraVal);

        ProductKeyEncoder keyi;
        string err;
        if(!NewProductKeyEncoder(group, serial, security, upgradeVal, checksumVal, extraVal, keyi, err)){
            printf("Error: %s\n", err.c_str());
            return 1;
        }

        printf("%s\n", keyi.Key5x5.c_str());

    } else {
        printf("Unknown command. Use: encode, decode, or template\n");
        return 1;
    }

    return 0;
}